src/extra/components/_adq.pyx

```diff
@@ -34,9 +41,15 @@ def _reshape_flat_pulses(
 
         prev_pid = cur_pid
         start = first_pulse_offset + (cur_pid - pid_offset) * clock_ratio
-        end = start + samples_per_pulse
 
-        if end > trace_len:
-            raise ValueError(f'trace too short for pulse at {start}:{end}')
+        # The end of this pulse is beyond the trace length, try to
+        # copy what remains and fill with placeholder as needed.
+        samples_to_copy = min(max(trace_len - start, 0), samples_per_pulse)
+
+        # Copy if data available.
+        memcpy(&out[i, 0], &traces[j, start],
+               samples_to_copy * sizeof(traces[0, 0]))
 
-        memcpy(&out[i, 0], &traces[j, start], bytes_per_pulse)
+        # Fill in the rest of this pulse with the placeholder value.
+        for k in range(samples_to_copy, samples_per_pulse):
+            out[i, k] = placeholder_val
```

Comment:
`memset` is fill memory with `0xff`, which is decoded as `-1` or `nan` depending on type. You may simplify `nan` placeholder value selection and padding:
```suggestion
        memset(&out[i, samples_to_copy], -1, (samples_per_pulse - samples_to_copy) * sizeof(out[0, 0]))
```

Reply:
Some googling suggests that while this will work most of the time, it's in the realm of undefined behaviour. So I'd rather not.